src/main.rs

```diff
@@ -414,4 +415,20 @@ async fn start_component(
 		)
 		.await
 		.unwrap_or_else(|_| panic!("failed to start {}", cmd));
+	#[cfg(feature = "systemd")]
+	{
+		if *is_systemd_used() {
+			//currently pid is optional hence the double unwrap
+			let pids = process_manager.get_pid(key).await.unwrap().unwrap();
```

Comment:
So does this crash, if we don't have a pid?

Reply:
It wouldn't crash in that case, just spawn a scope that manages nothing. There is another systemd function to set a PIDs for a scope after the fact. 

I'm not sure there is a situation where it would be `None`. I made the value optional for two reasons. The first being, it's optional for [tokio child](https://docs.rs/tokio/latest/tokio/process/struct.Child.html#method.id), the second being I had to set it after processData has been created. 

There might be a situation where the PID would be incorrect though. If a child is spawned, and immedately restarts, I don't think it would reuse the same pid.

Reply:
> I'm not sure there is a situation where it would be None. I made the value optional for two reasons.

But unwrapping a `None` does panic. We shouldn't assume an api always gives us `Some`, if it's signature says otherwise.

Reply:
> However we should make sure to destroy any created scopes, when we detect that the child has exited later. That would at least prevent any process re-using the PID to remain attached to the wrong scope for long.

So I guessing the best path is to map commands to scopes, and assume the pid changed, the scope must be altered. 

So `Dashmap`? `Arc<RwLock<HashMap>>`?

Reply:
Also, for anyone like me, who was curious. Restarting a failing process is not the responsibility of scopes, at least according to [this doc](https://www.freedesktop.org/software/systemd/man/latest/systemd.scope.html#:~:text=processes%20managed%20as%20scope%20units)

Reply:
Though this information is local to `cosmic-session`, I kinda dislike, that we already have a list of processes + metadata in the `process_manager` from `launchpad` and are now building our own list for additional metadata (the scope).

Maybe this isn't the best approach. Maybe we could just pass an optional scope to `process_manager.start` and let `launchpad` handle all of this setup. Including destroying the scope, when the process exits.

Alternatively we could allow attaching some arbitrary data to a process launched, which could then be passed to the closures in `with_on_start` and `with_on_exit`.

cc @wash2, any thoughts on this?

Reply:
> Restarting a failing process is not the responsibility of scopes

Right, it is launch-pads, so our responsibility. So we want to properly cleanup, before restarting.

Reply:
> Maybe this isn't the best approach. Maybe we could just pass an optional scope to `process_manager.start` and let `launchpad` handle all of this setup. Including destroying the scope, when the process exits.
>
> Alternatively we could allow attaching some arbitrary data to a process launched, which could then be passed to the closures in `with_on_start` and `with_on_exit`. Then we also don't need to depend (even if optional) on systemd in launch-pad.

I'm fine with either. I can recreate the zbus interface stuff in launch-pad. 

It's probably best to keep the option open in either case to have scopes capable of handling more than one PID, and whether it's here or in `launch-pad` there are a lot more capabilities of scopes we might want to make use of, such as cpu or memory limit, or things like on oom, which get will likely be triggered quite a bit if vscode + chrome + cargo or some other build system are active at the same time. maybe a `ScopeConfig` struct?

Reply:
I don't know, if we actually want to force these values via the programmable interface.

If they have stable names, these options can be set via drop-in config files, which is more flexible to customize for users and/or distros.

Reply:
> Alternatively we could allow attaching some arbitrary data to a process launched, which could then be passed to the closures in with_on_start and with_on_exit.
Then we also don't need to depend (even if optional) on systemd in launch-pad.

Ya, if most of the features are better configured through config files, this might be better for now.

Reply:
Currently trying to move systemd scope spawning to `with_start`, and I'm running into issues moving a clone of command. Process Manager currently doesn't expose a function to get the name of a command or a getter for processdata. It also seems I can't access private methods from within the closure so I can't just directly access inner. In order to pass cmd, I need to capture the evironment (`async move ||`) but I get the error/warning async closures are currently unstable. I'm guessing that switching to unstable is probably not the desired solution.  
<details>
```rust
.with_on_start(move |pman, pkey, will_restart| {
	#[cfg(feature = "systemd")]
	async move || {
		if *is_systemd_used() {
			//currently pid is optional hence the double unwrap
			if let Some(pids) = pman.get_pid(pkey).await.unwrap_or_else(|err| {
				warn!(
					"Failed to get pid for {}. Process manager failed with {}",
					cmd_clone2.clone(),
					err
				);
				None
			}) {
				//spawn_scope takes a vec of pids in case we want to spawn a scope for multiple processes
				if let Err(err) = spawn_scope(cmd_clone2.clone(), vec![pids]).await
				{
					warn!(
		"Failed to spawn scope for {}. Creating transient unit failed with {}",
		cmd_clone2.clone(), err
	);
				};
			}
		}
	}
})
```
</details>

I could open a PR to expose cmd name in with start, but if the problem is moving state, then how would deserializing scope configuration work without moving the systemd stuff to launch-pad? 

slightly unrelated, but is lauch-pad used by anything else, such as pop/cosmic-launcher?

Reply:
> slightly unrelated, but is lauch-pad used by anything else, such as pop/cosmic-launcher?

Ya, it is used by cosmic-panel for spawning applets.

> In order to pass cmd, I need to capture the evironment (async move ||) but I get the error/warning async closures are currently unstable. I'm guessing that switching to unstable is probably not the desired solution.

Why do you need `async move || {...}`? Could `move || async move {...}`, or something like that, work for your purposes? `start_component` moves the logging spans into the closure, but you might want to call `to_string` so that you don't run into lifetime issues so that you could do it the same way. Or you could make the argument `cmd: impl Into<Cow<'static, str>>` and convert it to `Cow<'static, str>` right away, so it can be moved into the closures just like the spans. Hopefully I've not misunderstood the issue you are facing here :sweat_smile: 

That being said, I also wouldn't be against a PR that exposes the name from launchpad.

Reply:
I did an org wide search for launch-pad to see what was using it. Right now it's used in `cosmic-panel`

> > slightly unrelated, but is lauch-pad used by anything else, such as pop/cosmic-launcher?
> 
> Ya, it is used by cosmic-panel for spawning applets.
> 
> > In order to pass cmd, I need to capture the evironment (async move ||) but I get the error/warning async closures are currently unstable. I'm guessing that switching to unstable is probably not the desired solution.
> 
> Why do you need `async move || {...}`? 

I mainly need some way to pass in the command, which I'm using to generate the scope name. so long as we are using a scope per process, that's really the only extra state we need and we can get away with just adding a function call launch-pad side. 

> Could `move || async move {...}`, or something like that, work for your purposes? 

I was not even aware that was a thing, but I tried that (along with the change of cmd to the mentioned typed signature, and I get 
```
`{closure@src/main.rs:384:7: 384:14}` is not a future
the trait `futures_util::Future` is not implemented for closure `{closure@src/main.rs:384:7: 384:14}`
main.rs(384, 7): consider calling this closure
process.rs(113, 6): required by a bound in `launch_pad::process::Process::with_on_start`
```
might be possible to change the function signature of the arg to startwith, but I'm not sure if that's going to break anything.

> That being said, I also wouldn't be against a PR that exposes the name from launchpad.

So I tried using the `get_exe_and_pid` function I added and I'm still having issues related to incompatible types. It's saying sync is not implemented for `dyn futures_util::Future<Output = Result<conn::handshake::Authenticated, zbus::Error>> + std::marker::Send` which (after inlining my function) I think is coming from `zbus::Connection::session()`. 

I'll poke around with this more tomorrow

Reply:
Ok, feel free to push what you have for it and ping if you'd like me to take a look too :+1:

Reply:
@wash2 yeah, kind of stuck on this one. Within the function, I can comment out all the code except the call to `Connection::session()` and the compiler message is still present, so this might be something that is currently missing from zbus. I'm uploading the full compiler diagnostic if it helps
[diagnostics.txt](https://github.com/user-attachments/files/15991367/diagnostics.txt)

Reply:
Wow, I guess I should learn to read diagnostics better. It turns out that this was adding the feature tokio to zbus

Reply:
Ok, thanks :slightly_smiling_face:

Reply:
hey when you get back around to this. I got some helpful information from the zbus maintainers ([see zbus#869](https://github.com/dbus2/zbus/issues/869)). It turns out it's the future returned by connection (and it seems any awaitable Result). 


I've managed to work around the issue by using `tokio::task::spawn_local`, I still need to test the changes though

Reply:
https://github.com/pop-os/launch-pad/pull/14 should fix the issue you are facing, or at least did when I tested it earlier.

Reply:
It did, I tested it last night. I guess once that is merged, I can go ahead and resolve this.

Reply:
@skewballfox I merged https://github.com/pop-os/launch-pad/pull/14, so the Sync bound should be resolved now by updating launch-pad